libnet/src/pf_key.rs

```diff
@@ -0,0 +1,830 @@
+        let mtype = if update {
+            MessageType::Update
+        } else {
+            MessageType::Add
+        };
+
+        Self {
+            header: Header::new(mtype, SaType::TcpSig, size_of::<Self>()),
+            association: Association::default(),
+            lifetime: Lifetime::hard(valid_time, valid_time),
+            src: Address::src(src, IPPROTO_TCP as u8),
+            dst: Address::dst(dst, IPPROTO_TCP as u8),
+            key: StrAuth::new(authstring),
+        }
+    }
+}
+
+/// A packet to delete a TCP-MD5 security association.
+#[repr(C, packed)]
+pub struct TcpMd5DeleteKeyRequest {
+    /// Packet header.
+    pub header: Header,
+    /// Association info.
+    pub association: Association,
+    /// Source socket address to unbind.
+    pub src: Address,
+    /// Destination socket address to unbind.
+    pub dst: Address,
+}
+
+impl TcpMd5DeleteKeyRequest {
+    /// Create a new TCP-MD5 delete key request.
+    pub fn new(src: SockAddr, dst: SockAddr) -> Self {
+        Self {
+            header: Header::new(
+                MessageType::Delete,
+                SaType::TcpSig,
+                size_of::<Self>(),
+            ),
+            association: Association::default(),
+            src: Address::src(src, IPPROTO_TCP as u8),
+            dst: Address::dst(dst, IPPROTO_TCP as u8),
+        }
+    }
+}
+
+/// A packet to request info about a TCP-MD5 security association.
+#[repr(C, packed)]
+pub struct TcpMd5GetKeyRequest {
+    /// Packet header.
+    pub header: Header,
+    /// Association info.
+    pub association: Association,
+    /// Source socket address predicate.
+    pub src: Address,
+    /// Destination socket address predicate.
+    pub dst: Address,
+}
+
+impl TcpMd5GetKeyRequest {
+    /// Create a new TCP-MD5 get key request.
+    pub fn new(src: SockAddr, dst: SockAddr) -> Self {
+        Self {
+            header: Header::new(
+                MessageType::Get,
+                SaType::TcpSig,
+                size_of::<Self>(),
+            ),
+            association: Association::default(),
+            src: Address::src(src, IPPROTO_TCP as u8),
+            dst: Address::dst(dst, IPPROTO_TCP as u8),
+        }
+    }
+}
+
+/// Response information returned from kernel from a key association request.
+#[derive(Debug)]
+pub struct GetAssociationResponse {
+    pub header: Header,
+    pub extensions: Vec<Extension>,
+}
+
+/// Add a TCP-MD5 security association for the provided source and destination
+/// address with `authstring` as the key that is valid for `valid_time` after
+/// creation.
+pub fn tcp_md5_key_add(
+    src: SockAddr,
+    dst: SockAddr,
+    authstring: &str,
+    valid_time: Duration,
+) -> Result<(), Error> {
+    tcp_md5_key_set(src, dst, authstring, valid_time, false)
+}
+
+/// Update a TCP-MD5 security association for the provided source and destination
+/// address with `authstring` as the key that is valid for `valid_time` after
+/// creation.
```

Comment:
This does imply that the key could be changed in an update message, which it can't.

Reply:
This has been changed.